### PR TITLE
Add Testing RSpec for HW02 and HW03

### PR DESCRIPTION
**Pull request by Megonar**

### CODE REVIEW

File: HW02/Ramanovich_Dzmitry/my_array_method_spec.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+require_relative 'my_array_method'
+
+using MyArrayMethod
+
+RSpec.describe MyArrayMethod do
+  let(:array_empty) { [] }
```
Comment (labs-ruby) [inline HW02/Ramanovich_Dzmitry/my_array_method_spec.rb:8]: you do not need this variable, mutate `array` in each context instead

File: HW03/Ramanovich_Dzmitry/task_1_spec.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'rspec'
+require_relative 'task_1'
+
+RSpec.describe 'Test task_1.rb' do
+  let(:logs) do
```
Comment (labs-ruby) [inline HW03/Ramanovich_Dzmitry/task_1_spec.rb:7]: please define variables in each context instead defining them globally

File: HW02/Ramanovich_Dzmitry/my_array_method_spec.rb (hunk 1 above)
Comment (Megonar) [reply HW02/Ramanovich_Dzmitry/my_array_method_spec.rb]: Done. I`m deleted variable MeArrayMethod

File: HW03/Ramanovich_Dzmitry/task_1_spec.rb (hunk 2 above)
Comment (Megonar) [reply HW03/Ramanovich_Dzmitry/task_1_spec.rb:7]: Done. I`m created variables in each context

File: HW02/Ramanovich_Dzmitry/my_array_method_spec.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      end
+    end
+
+    context 'when array is not empty' do
+      let(:array) { [1, 2, 3] }
+      it 'returns the same number of elements as the original array' do
+        expect(array.my_map(&:even?).count).to eq(array.count)
+      end
+
+      it 'origin array does not change' do
```
Comment (B0RlS) [inline HW02/Ramanovich_Dzmitry/my_array_method_spec.rb:22]:
```suggestion
      it 'does not modify the original array' do
```

File: HW03/Ramanovich_Dzmitry/task_1_spec.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    let(:string_with_errors) do
+      '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>'
+    end
+    it 'return first line with errors' do
+      expect(task_1(logs_with_errors)).to eq(string_with_errors)
+    end
+  end
+
+  context 'when no arguments is given' do
+    it 'return Error' do
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_1_spec.rb:58]:
```suggestion
    it 'returns Error' do
```

File: HW02/Ramanovich_Dzmitry/my_array_method_spec.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        expect(array.my_each { |x| x + 2 }).to eq([1, 2, 3, 4, 5])
+      end
+    end
+  end
+
+  describe '#my_select' do
+    context 'when array is empty' do
+      let(:array_empty) { [] }
+
+      it 'return eampty array' do
```
Comment (B0RlS) [inline HW02/Ramanovich_Dzmitry/my_array_method_spec.rb:67]:
```suggestion
      it 'returns eampty array' do
```

File: HW02/Ramanovich_Dzmitry/my_array_method_spec.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        expect(array.my_map { |x| x * 2 }).to eq([2, 4, 6, 8, 10])
+      end
+    end
+  end
+
+  describe '#my_each' do
+    context 'when array is empty' do
+      let(:array_empty) { [] }
+
+      it 'return eampty array' do
```
Comment (B0RlS) [inline HW02/Ramanovich_Dzmitry/my_array_method_spec.rb:41]:
```suggestion
      it 'returns eampty array' do
```

File: HW02/Ramanovich_Dzmitry/my_array_method_spec.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        expect(array_empty.my_select { |x| x > 1 }).to eq([])
+      end
+    end
+
+    context 'when an array in input block' do
+      let(:array) { [1, 2, 3, 4, 5] }
+
+      it 'Returns a new array containing all elements of ary for which the given block returns a true value' do
```
Comment (B0RlS) [inline HW02/Ramanovich_Dzmitry/my_array_method_spec.rb:75]:
```suggestion
      it 'returnы a new array with a suitable condition' do
```

File: HW03/Ramanovich_Dzmitry/task_1_spec.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+        2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+        10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+      LOG
+    end
+
+    let(:string_with_errors) do
+      '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>'
+    end
+    it 'return full text first line with an error' do
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_1_spec.rb:20]: Don't forget the 's'

Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_1_spec.rb:20]: Don't forget the 's'

File: HW03/Ramanovich_Dzmitry/task_1_spec.rb [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+      <<~LOG
+        10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+        10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+        2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+        10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+        2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+      LOG
+    end
+
+    let(:string_with_errors) do
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_1_spec.rb:49]:
```suggestion
    let(:expected_results) do
```

File: HW03/Ramanovich_Dzmitry/task_3.rb [hunk 10]
```diff
@@ -1,15 +1,19 @@
...
-
-def method(str)
-  str.each_cons(2).map { |str_one, str_two| Time.parse(str_two) - Time.parse(str_one) }.join(',')
+  if str.empty?
+    '0'
+  else
+    subtraction << str.each_cons(2).map do |str_one, str_two|
+      Time.parse(str_two) - Time.parse(str_one)
+    end.join(',')
+  end
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_3.rb:18]: move it to private method

File: HW03/Ramanovich_Dzmitry/task_3_spec.rb [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    end
+
+    context 'when zero events' do
+      it 'return 0' do
+        expect(task_3('')).to eq('0')
+      end
+    end
+
+    context 'when much events' do
+      let(:log_much_events) do
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_3_spec.rb:33]:
```suggestion
      let(:log) do
```

File: HW03/Ramanovich_Dzmitry/task_3_spec.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+      end
+    end
+
+    context 'when zero events' do
+      it 'return 0' do
+        expect(task_3('')).to eq('0')
+      end
+    end
+
+    context 'when much events' do
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_3_spec.rb:32]:
```suggestion
    context 'when log has more then two events' do
```

File: HW03/Ramanovich_Dzmitry/task_4_spec.rb [hunk 13]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+require 'rspec'
+require_relative 'task_4'
+
+RSpec.describe 'Test task_4' do
+  context 'when input only digits' do
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_4_spec.rb:7]:
```suggestion
  context 'when input has only digits' do
```

File: HW03/Ramanovich_Dzmitry/task_3.rb (hunk 10 above)
Comment (Megonar) [reply HW03/Ramanovich_Dzmitry/task_3.rb]: Dona

File: HW03/Ramanovich_Dzmitry/task_1_spec.rb (hunk 8 above)
Comment (Megonar) [reply HW03/Ramanovich_Dzmitry/task_1_spec.rb]: Done

Comment (Megonar) [reply HW03/Ramanovich_Dzmitry/task_1_spec.rb]: Done

File: HW03/Ramanovich_Dzmitry/task_3.rb [hunk 14]
```diff
@@ -7,9 +7,18 @@ def task_3(logs)
...
+  decision(str)
 end
 
-def method(str)
-  str.each_cons(2).map { |str_one, str_two| Time.parse(str_two) - Time.parse(str_one) }.join(',')
+private
+
+def decision(str)
+  subtraction = []
+  subtraction << if str.empty?
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_3.rb:17]: Move this check to private method please

File: HW03/Ramanovich_Dzmitry/task_3.rb [hunk 15]
```diff
@@ -7,9 +7,18 @@ def task_3(logs)
...
+                   '0'
+                 else
+                   str.each_cons(2).map do |str_one, str_two|
```
Comment (B0RlS) [inline HW03/Ramanovich_Dzmitry/task_3.rb:20]: And this iterator

File: HW03/Ramanovich_Dzmitry/task_3.rb (hunk 14 above)
Comment (Megonar) [reply HW03/Ramanovich_Dzmitry/task_3.rb]: Done

File: HW03/Ramanovich_Dzmitry/task_3.rb (hunk 15 above)
Comment (Megonar) [reply HW03/Ramanovich_Dzmitry/task_3.rb]: Done

**Pull request merged**